embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/scheduling/SchedulingServiceImpl.kt

```diff
@@ -1,14 +1,162 @@
+@file:Suppress("FunctionOnlyReturningConstant")
+
 package io.embrace.android.embracesdk.internal.delivery.scheduling
 
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.internal.comms.api.ApiResponse
+import io.embrace.android.embracesdk.internal.comms.api.Endpoint
 import io.embrace.android.embracesdk.internal.delivery.execution.RequestExecutionService
+import io.embrace.android.embracesdk.internal.storage.PayloadReference
+import io.embrace.android.embracesdk.internal.storage.StorageService2
+import io.embrace.android.embracesdk.internal.worker.BackgroundWorker
+import java.io.InputStream
+import java.util.Collections
+import java.util.LinkedList
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
 
 internal class SchedulingServiceImpl(
```

Comment:
The general logic is as follows:

- When `onPayloadIntake()` is invoked, schedule a job on the `scheduleWorker` (using the non-IO background worker) to loop through the available payloads to be sent and attempts to send them if we aren't already running such a loop

- The loop will query for all the available payloads, and as long as a payload is eligible to be sent, queue it up to be sent

- When the loop finishes, it checks to see if there are additional payloads to send since the last query. Any payload that's in progress will be skipped (the current implementation can probably be optimized to do better filtering). The loop ends if it doesn't find any new payloads, and if there are payloads that ought to be retried, the earliest time will be used for a job to be scheduled to run then to check for payloads to deliver. The loop, at that point, should find at least that one payload and try to send it.

- When a payload is sent successfully, it is deleted via the storage service. 

- If it's not successful, it will be added to a Map of payloads that will be retried, with some metadata about when it should next be attempted, and how many failures there had been. The latter will be used to put a cap on how many retries are done (to be implemented). If a payload send retry is successful, this payload is removed. If it fails again, the metadata is updated.

- To facilitate this, `PayloadReference` was added as an abstraction of the cached payload file. A concept like this is needed so we don't just use the `File` reference, which is unstable and could be duplicated. The attributes in this data class should uniquely identify a payload on disk to be sent, and also provide enough metadata required for various parts of the delivery process.